src/mscorlib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5994,7 +5994,10 @@ public static Task WhenAll(IEnumerable<Task> tasks)
             foreach (Task task in tasks)
             {
                 if (task == null) ThrowHelper.ThrowArgumentException(ExceptionResource.Task_MultiTaskContinuation_NullTask, ExceptionArgument.tasks);
-                taskList.Add(task);
+                if (!task.IsRanToCompletion) // no need to wait on the task if it's already successfully completed
```

Comment:
In case the links didn't work, I was referring to lines 6107 and 6142 in the modified file.